### PR TITLE
SONARJAVA-2446 do not report exceptional flow if there is non-exceptional ones

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/se/Flow.java [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+  }
+
+  @Override
+  public List<JavaFileScannerContext.Location> subList(int fromIndex, int toIndex) {
+    return items.subList(fromIndex, toIndex);
+  }
+
+  public static class Builder {
+
+    private final ImmutableList.Builder<JavaFileScannerContext.Location> builder;
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/Flow.java]: MAJOR Rename field "builder" [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1700)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: java-frontend/src/main/java/org/sonar/java/se/Flow.java [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+  }
+
+  @Override
+  public ListIterator<JavaFileScannerContext.Location> listIterator(int index) {
+    return items.listIterator(index);
+  }
+
+  @Override
+  public List<JavaFileScannerContext.Location> subList(int fromIndex, int toIndex) {
+    return items.subList(fromIndex, toIndex);
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/Flow.java]: I'm not convinced by this one. Probably needs to be instead:
```
return new Flow(items.subList(fromIndex, toIndex), exceptional);
```

File: java-frontend/src/main/java/org/sonar/java/se/checks/SECheck.java [hunk 3]
```diff
@@ -103,7 +107,12 @@ public String getMessage() {
     }
 
     public Set<List<JavaFileScannerContext.Location>> getFlows() {
-      return flows;
+      Stream<Flow> flowsAsStream = flows.stream();
```
Comment (saberduck) [inline java-frontend/src/main/java/org/sonar/java/se/checks/SECheck.java]: Maybe bit more clearer like this
```
Set<List<JavaFileScannerContext.Location>> nonExceptionalFlows = flows.stream().filter(Flow::isNonExceptional).map(Flow::elements).collect(Collectors.toSet());
if (!nonExceptionalFlows.isEmpty()) {
    return nonExceptionalFlows;
}
return flows.stream().map(Flow::elements).collect(Collectors.toSet());
```
Also you can drop `isExceptional` method on `Flow` afterwards

**Pull request merged**